---
title: "Data_Setup"
output: html_notebook
---

# Goal: Make data amenable to use within the application

# Install necessary packages
```{r}
#install.packages("stringr")
library(stringr)

#if (!require("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")
#BiocManager::install(version = "3.15")

#BiocManager::install("EnrichmentBrowser")
library(EnrichmentBrowser)
```

# 1. Supported_Species: loads in metadata on species and strains that are included in the application 
```{r}
Supported_Species <- read.csv("./App_Data/Supported_Species.csv")
View(Supported_Species)
```

# 2. Target_Prediction_Data: loads in IntaRNA data tables for each species
```{r}
Target_Prediction_Data <- vector(mode = "list", length = nrow(Supported_Species))

Target_Prediction_Data[[1]] <- read.csv("./App_Data/Energy_Tables/AF_Energy.csv")
Target_Prediction_Data[[2]] <- read.csv("./App_Data/Energy_Tables/BC_Energy.csv")
Target_Prediction_Data[[3]] <- read.csv("./App_Data/Energy_Tables/BF_Energy.csv")
Target_Prediction_Data[[4]] <- read.csv("./App_Data/Energy_Tables/CA_Energy.csv")
Target_Prediction_Data[[5]] <- read.csv("./App_Data/Energy_Tables/CD_Energy.csv")
Target_Prediction_Data[[6]] <- read.csv("./App_Data/Energy_Tables/CP_Energy.csv")
Target_Prediction_Data[[7]] <- read.csv("./App_Data/Energy_Tables/EC_Energy.csv")
Target_Prediction_Data[[8]] <- read.csv("./App_Data/Energy_Tables/HI_Energy.csv")
Target_Prediction_Data[[9]] <- read.csv("./App_Data/Energy_Tables/MAv_Energy.csv")
Target_Prediction_Data[[10]] <- read.csv("./App_Data/Energy_Tables/PA_Energy.csv")
Target_Prediction_Data[[11]] <- read.csv("./App_Data/Energy_Tables/PM_Energy.csv")
Target_Prediction_Data[[12]] <- read.csv("./App_Data/Energy_Tables/SA_Energy.csv")
Target_Prediction_Data[[13]] <- read.csv("./App_Data/Energy_Tables/SM_Energy.csv")
Target_Prediction_Data[[14]] <- read.csv("./App_Data/Energy_Tables/SP2_Energy.csv")
Target_Prediction_Data[[15]] <- read.csv("./App_Data/Energy_Tables/SP_Energy.csv")
Target_Prediction_Data[[16]] <- read.csv("./App_Data/Energy_Tables/SS2_Energy.csv")
Target_Prediction_Data[[17]] <- read.csv("./App_Data/Energy_Tables/SS_Energy.csv")
Target_Prediction_Data[[18]] <- read.csv("./App_Data/Energy_Tables/AX_Energy.csv")
Target_Prediction_Data[[19]] <- read.csv("./App_Data/Energy_Tables/BM_Energy.csv")
Target_Prediction_Data[[20]] <- read.csv("./App_Data/Energy_Tables/MA_Energy.csv")

names(Target_Prediction_Data) <- Supported_Species$Species

for (i in 1:length(Target_Prediction_Data)) {
  rownames(Target_Prediction_Data[[i]]) <- Target_Prediction_Data[[i]][,1]
  Target_Prediction_Data[[i]] <- Target_Prediction_Data[[i]][,-1]
} #Moves first column into row names (will make analysis easier in app)
  
View(Target_Prediction_Data)
```

# 3 Gene_Annotation_Data: Create gene annotation data table for each species
```{r}
Gene_Annotation_Data <- vector(mode = "list", length = nrow(Supported_Species))

#First need to take the gene identifier column for each IntaRNA data table, and use RegExp to extract out just the ID that we want to map to KEGG IDs. The NCBI Gene ID is what we will use for the initial mapping. And if the gene ID is not in the data, use the protein ID (For protein IDs, have to 'filter by taxonomy' at top left)

{
  #A. fumigatus
  {
    Target_Prediction_Data[[1]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[1]]), "GeneID:\\s*[0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[1]]$Gene_ID)) {
        Target_Prediction_Data[[1]]$Gene_ID[i] <- substr(Target_Prediction_Data[[1]]$Gene_ID[i], 8, nchar(Target_Prediction_Data[[1]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[1]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/AFGeneIDs.csv")
  }
  
  #B. cenocepacia
  {
    Target_Prediction_Data[[2]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[2]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[2]]$Gene_ID)) {
        Target_Prediction_Data[[2]]$Gene_ID[i] <- substr(Target_Prediction_Data[[2]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[2]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[2]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/BCGeneIDs.csv")
  }
  
  #B. fragilis
  {
    Target_Prediction_Data[[3]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[3]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[3]]$Gene_ID)) {
        Target_Prediction_Data[[3]]$Gene_ID[i] <- substr(Target_Prediction_Data[[3]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[3]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[3]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/BFGeneIDs.csv")
  }
  
  #C. albicans
  {
    Target_Prediction_Data[[4]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[4]]), "GeneID:\\s*[0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[4]]$Gene_ID)) {
        Target_Prediction_Data[[4]]$Gene_ID[i] <- substr(Target_Prediction_Data[[4]]$Gene_ID[i], 8, nchar(Target_Prediction_Data[[4]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[4]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/CAGeneIDs.csv")
  }
  
  #C. difficile
  {
    Target_Prediction_Data[[5]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[5]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[5]]$Gene_ID)) {
        Target_Prediction_Data[[5]]$Gene_ID[i] <- substr(Target_Prediction_Data[[5]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[5]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[5]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/CDGeneIDs.csv")
  }
  
  #C. perfringens
  {
    Target_Prediction_Data[[6]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[6]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[6]]$Gene_ID)) {
        Target_Prediction_Data[[6]]$Gene_ID[i] <- substr(Target_Prediction_Data[[6]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[6]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[6]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/CPGeneIDs.csv")
  }
  
  #E. coli
  {
    Target_Prediction_Data[[7]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[7]]), "protein_id=\\s*[WNYP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[7]]$Gene_ID)) {
        Target_Prediction_Data[[7]]$Gene_ID[i] <- substr(Target_Prediction_Data[[7]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[7]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[7]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/ECGeneIDs.csv")
  }
  
  #H. influenza
  {
    Target_Prediction_Data[[8]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[8]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[8]]$Gene_ID)) {
        Target_Prediction_Data[[8]]$Gene_ID[i] <- substr(Target_Prediction_Data[[8]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[8]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[8]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/HIGeneIDs.csv")
  }
  
  #M. avium
  {
    Target_Prediction_Data[[9]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[9]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[9]]$Gene_ID)) {
        Target_Prediction_Data[[9]]$Gene_ID[i] <- substr(Target_Prediction_Data[[9]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[9]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[9]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/MAvGeneIDs.csv")
  }
  
  #P. aeruginosa
  {
    Target_Prediction_Data[[10]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[10]]), "GeneID:\\s*[0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[10]]$Gene_ID)) {
        Target_Prediction_Data[[10]]$Gene_ID[i] <- substr(Target_Prediction_Data[[10]]$Gene_ID[i], 8, nchar(Target_Prediction_Data[[10]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[10]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/PAGeneIDs.csv")
  }
  
  #P. melaninogenica
  {
    Target_Prediction_Data[[11]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[11]]), "GeneID:\\s*[0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[11]]$Gene_ID)) {
        Target_Prediction_Data[[11]]$Gene_ID[i] <- substr(Target_Prediction_Data[[11]]$Gene_ID[i], 8, nchar(Target_Prediction_Data[[11]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[11]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/PMGeneIDs.csv")
  }
  
  #S. aureus
  {
    Target_Prediction_Data[[12]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[12]]), "GeneID:\\s*[0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[12]]$Gene_ID)) {
        Target_Prediction_Data[[12]]$Gene_ID[i] <- substr(Target_Prediction_Data[[12]]$Gene_ID[i], 8, nchar(Target_Prediction_Data[[12]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[12]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/SAGeneIDs.csv")
  }
  
  #S. maltophilia
  {
    Target_Prediction_Data[[13]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[13]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[13]]$Gene_ID)) {
        Target_Prediction_Data[[13]]$Gene_ID[i] <- substr(Target_Prediction_Data[[13]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[13]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[13]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/SMGeneIDs.csv")
  }
  
  #S. parasanguinis
  {
    Target_Prediction_Data[[14]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[14]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[14]]$Gene_ID)) {
        Target_Prediction_Data[[14]]$Gene_ID[i] <- substr(Target_Prediction_Data[[14]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[14]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[14]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/SPaGeneIDs.csv")
  }
  
  #S. pneumonia
  {
    Target_Prediction_Data[[15]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[15]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[15]]$Gene_ID)) {
        Target_Prediction_Data[[15]]$Gene_ID[i] <- substr(Target_Prediction_Data[[15]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[15]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[15]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/SPGeneIDs.csv")
  }
  
  #S. salivarius
  {
    Target_Prediction_Data[[16]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[16]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[16]]$Gene_ID)) {
        Target_Prediction_Data[[16]]$Gene_ID[i] <- substr(Target_Prediction_Data[[16]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[16]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[16]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/SSalGeneIDs.csv")
  }
  
  #S. sanguinus
  {
    Target_Prediction_Data[[17]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[17]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[17]]$Gene_ID)) {
        Target_Prediction_Data[[17]]$Gene_ID[i] <- substr(Target_Prediction_Data[[17]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[17]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[17]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/SSGeneIDs.csv")
  }
  
  #A. xylosoxidans
  {
    Target_Prediction_Data[[18]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[18]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[18]]$Gene_ID)) {
        Target_Prediction_Data[[18]]$Gene_ID[i] <- substr(Target_Prediction_Data[[18]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[18]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[18]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/AXGeneIDs.csv")
  }
  
  #B. multivorans
  {
    Target_Prediction_Data[[19]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[19]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[19]]$Gene_ID)) {
        Target_Prediction_Data[[19]]$Gene_ID[i] <- substr(Target_Prediction_Data[[19]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[19]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[19]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/BMGeneIDs.csv")
  }
  
  #M. abscessus
  {
    Target_Prediction_Data[[20]]$Gene_ID <- unlist(str_extract(rownames(Target_Prediction_Data[[20]]), "protein_id=\\s*[WP_.0-9]+\\s*"))
    for (i in 1:length(Target_Prediction_Data[[20]]$Gene_ID)) {
        Target_Prediction_Data[[20]]$Gene_ID[i] <- substr(Target_Prediction_Data[[20]]$Gene_ID[i], 12, nchar(Target_Prediction_Data[[20]]$Gene_ID[i]))
    } 
    write.csv(Target_Prediction_Data[[20]]$Gene_ID, "./App_Data/Annotation_Data/Gene_IDs/MAGeneIDs.csv")
  }
  
  
}

#Once all the NCBI Gene IDs/RefSeq Protein IDs are gathered, convert to Uniprot IDs using the Uniprot ID Mapping Tool: https://www.uniprot.org/id-mapping. Then use the same tool to convert Uniprot IDs to KEGG IDs. The resulting annotation tables are stored in the App_Data folder. 

#Load all the annotation data into the Gene_Annotation_Data List (same order of species/strains as Target_Prediction_Data object)
{
Gene_Annotation_Data[[1]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/AF_Annotations.csv")
Gene_Annotation_Data[[2]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/BC_Annotations.csv")
Gene_Annotation_Data[[3]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/BF_Annotations.csv")
Gene_Annotation_Data[[4]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/CA_Annotations.csv")
Gene_Annotation_Data[[5]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/CD_Annotations.csv")
Gene_Annotation_Data[[6]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/CP_Annotations.csv")
Gene_Annotation_Data[[7]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/EC_Annotations.csv")
Gene_Annotation_Data[[8]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/HI_Annotations.csv")
Gene_Annotation_Data[[9]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/MAv_Annotations.csv")
Gene_Annotation_Data[[10]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/PA_Annotations.csv")
Gene_Annotation_Data[[11]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/PM_Annotations.csv")
Gene_Annotation_Data[[12]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/SA_Annotations.csv")
Gene_Annotation_Data[[13]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/SM_Annotations.csv")
Gene_Annotation_Data[[14]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/SPa_Annotations.csv")
Gene_Annotation_Data[[15]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/SP_Annotations.csv")
Gene_Annotation_Data[[16]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/SSal_Annotations.csv")
Gene_Annotation_Data[[17]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/SS_Annotations.csv")
Gene_Annotation_Data[[18]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/AX_Annotations.csv")
Gene_Annotation_Data[[19]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/BM_Annotations.csv")
Gene_Annotation_Data[[20]][[1]] <- read.csv("./App_Data/Annotation_Data/Annotation_Tables/MA_Annotations.csv")

Gene_Annotation_Data[[1]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/AF_Conversion.csv")
Gene_Annotation_Data[[2]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/BC_Conversion.csv")
Gene_Annotation_Data[[3]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/BF_Conversion.csv")
Gene_Annotation_Data[[4]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/CA_Conversion.csv")
Gene_Annotation_Data[[5]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/CD_Conversion.csv")
Gene_Annotation_Data[[6]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/CP_Conversion.csv")
Gene_Annotation_Data[[7]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/EC_Conversion.csv")
Gene_Annotation_Data[[8]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/HI_Conversion.csv")
Gene_Annotation_Data[[9]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/MAv_Conversion.csv")
Gene_Annotation_Data[[10]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/PA_Conversion.csv")
Gene_Annotation_Data[[11]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/PM_Conversion.csv")
Gene_Annotation_Data[[12]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/SA_Conversion.csv")
Gene_Annotation_Data[[13]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/SM_Conversion.csv")
Gene_Annotation_Data[[14]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/SPa_Conversion.csv")
Gene_Annotation_Data[[15]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/SP_Conversion.csv")
Gene_Annotation_Data[[16]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/SSal_Conversion.csv")
Gene_Annotation_Data[[17]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/SS_Conversion.csv")
Gene_Annotation_Data[[18]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/AX_Conversion.csv")
Gene_Annotation_Data[[19]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/BM_Conversion.csv")
Gene_Annotation_Data[[20]][[2]] <- read.csv("./App_Data/Annotation_Data/Uniprot_Conversion/MA_Conversion.csv")

Gene_Annotation_Data[[1]][[2]] <- Gene_Annotation_Data[[1]][[2]][,c(1:2)]
Gene_Annotation_Data[[2]][[2]] <- Gene_Annotation_Data[[2]][[2]][,c(1:2)]
Gene_Annotation_Data[[3]][[2]] <- Gene_Annotation_Data[[3]][[2]][,c(1:2)]
Gene_Annotation_Data[[4]][[2]] <- Gene_Annotation_Data[[4]][[2]][,c(1:2)]
Gene_Annotation_Data[[5]][[2]] <- Gene_Annotation_Data[[5]][[2]][,c(1:2)]
Gene_Annotation_Data[[6]][[2]] <- Gene_Annotation_Data[[6]][[2]][,c(1:2)]
Gene_Annotation_Data[[7]][[2]] <- Gene_Annotation_Data[[7]][[2]][,c(1:2)]
Gene_Annotation_Data[[8]][[2]] <- Gene_Annotation_Data[[8]][[2]][,c(1:2)]
Gene_Annotation_Data[[9]][[2]] <- Gene_Annotation_Data[[9]][[2]][,c(1:2)]
Gene_Annotation_Data[[10]][[2]] <- Gene_Annotation_Data[[10]][[2]][,c(1:2)]
Gene_Annotation_Data[[11]][[2]] <- Gene_Annotation_Data[[11]][[2]][,c(1:2)]
Gene_Annotation_Data[[12]][[2]] <- Gene_Annotation_Data[[12]][[2]][,c(1:2)]
Gene_Annotation_Data[[13]][[2]] <- Gene_Annotation_Data[[13]][[2]][,c(1:2)]
Gene_Annotation_Data[[14]][[2]] <- Gene_Annotation_Data[[14]][[2]][,c(1:2)]
Gene_Annotation_Data[[15]][[2]] <- Gene_Annotation_Data[[15]][[2]][,c(1:2)]
Gene_Annotation_Data[[16]][[2]] <- Gene_Annotation_Data[[16]][[2]][,c(1:2)]
Gene_Annotation_Data[[17]][[2]] <- Gene_Annotation_Data[[17]][[2]][,c(1:2)]
Gene_Annotation_Data[[18]][[2]] <- Gene_Annotation_Data[[18]][[2]][,c(1:2)]
Gene_Annotation_Data[[19]][[2]] <- Gene_Annotation_Data[[19]][[2]][,c(1:2)]
Gene_Annotation_Data[[20]][[2]] <- Gene_Annotation_Data[[20]][[2]][,c(1:2)]

View(Gene_Annotation_Data)

for (i in 1:length(Gene_Annotation_Data)) {
  names(Gene_Annotation_Data[[i]][[1]]) <- c("Uniprot_ID","KEGG_ID")
  names(Gene_Annotation_Data[[i]][[2]]) <- c("Gene_ID", "Uniprot_ID")
}

Gene_Annotation_Data[[1]][[3]] <- merge(Gene_Annotation_Data[[1]][[1]], Gene_Annotation_Data[[1]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[2]][[3]] <- merge(Gene_Annotation_Data[[2]][[1]], Gene_Annotation_Data[[2]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[3]][[3]] <- merge(Gene_Annotation_Data[[3]][[1]], Gene_Annotation_Data[[3]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[4]][[3]] <- merge(Gene_Annotation_Data[[4]][[1]], Gene_Annotation_Data[[4]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[5]][[3]] <- merge(Gene_Annotation_Data[[5]][[1]], Gene_Annotation_Data[[5]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[6]][[3]] <- merge(Gene_Annotation_Data[[6]][[1]], Gene_Annotation_Data[[6]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[7]][[3]] <- merge(Gene_Annotation_Data[[7]][[1]], Gene_Annotation_Data[[7]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[8]][[3]] <- merge(Gene_Annotation_Data[[8]][[1]], Gene_Annotation_Data[[8]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[9]][[3]] <- merge(Gene_Annotation_Data[[9]][[1]], Gene_Annotation_Data[[9]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[10]][[3]] <- merge(Gene_Annotation_Data[[10]][[1]], Gene_Annotation_Data[[10]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[11]][[3]] <- merge(Gene_Annotation_Data[[11]][[1]], Gene_Annotation_Data[[11]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[12]][[3]] <- merge(Gene_Annotation_Data[[12]][[1]], Gene_Annotation_Data[[12]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[13]][[3]] <- merge(Gene_Annotation_Data[[13]][[1]], Gene_Annotation_Data[[13]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[14]][[3]] <- merge(Gene_Annotation_Data[[14]][[1]], Gene_Annotation_Data[[14]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[15]][[3]] <- merge(Gene_Annotation_Data[[15]][[1]], Gene_Annotation_Data[[15]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[16]][[3]] <- merge(Gene_Annotation_Data[[16]][[1]], Gene_Annotation_Data[[16]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[17]][[3]] <- merge(Gene_Annotation_Data[[17]][[1]], Gene_Annotation_Data[[17]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[18]][[3]] <- merge(Gene_Annotation_Data[[18]][[1]], Gene_Annotation_Data[[18]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[19]][[3]] <- merge(Gene_Annotation_Data[[19]][[1]], Gene_Annotation_Data[[19]][[2]], by = "Uniprot_ID", all = TRUE)
Gene_Annotation_Data[[20]][[3]] <- merge(Gene_Annotation_Data[[20]][[1]], Gene_Annotation_Data[[20]][[2]], by = "Uniprot_ID", all = TRUE)
}

#Name the items in the list object appropriately
{
for (i in 1:length(Gene_Annotation_Data)) {
  names(Gene_Annotation_Data[[i]]) <- c("Uniprot-KEGG_Conversion", "GeneID-Uniprot_Conversion",   "Master_Annotation_File")
}

names(Gene_Annotation_Data) <- Supported_Species$Species
}

View(Gene_Annotation_Data)

```
# 4. KEGG_Gene_Lists: Assemble KEGG pathway gene lists
```{r}
KEGG_Gene_Lists <- vector(mode = "list", length = nrow(Supported_Species))

KEGG_Gene_Lists[[1]] <- getGenesets(org = "afm", db = "kegg")
KEGG_Gene_Lists[[2]] <- getGenesets(org = "bcj", db = "kegg")
KEGG_Gene_Lists[[3]] <- getGenesets(org = "bfs", db = "kegg")
KEGG_Gene_Lists[[4]] <- getGenesets(org = "cal", db = "kegg")
KEGG_Gene_Lists[[5]] <- getGenesets(org = "cdf", db = "kegg")
KEGG_Gene_Lists[[6]] <- getGenesets(org = "cpf", db = "kegg")
KEGG_Gene_Lists[[7]] <- getGenesets(org = "ecj", db = "kegg")
KEGG_Gene_Lists[[8]] <- getGenesets(org = "hiw", db = "kegg")
KEGG_Gene_Lists[[9]] <- getGenesets(org = "mav", db = "kegg")
KEGG_Gene_Lists[[10]] <- getGenesets(org = "pae", db = "kegg")
KEGG_Gene_Lists[[11]] <- getGenesets(org = "pmz", db = "kegg")
KEGG_Gene_Lists[[12]] <- getGenesets(org = "sao", db = "kegg")
KEGG_Gene_Lists[[13]] <- getGenesets(org = "sml", db = "kegg")
KEGG_Gene_Lists[[14]] <- getGenesets(org = "scp", db = "kegg")
KEGG_Gene_Lists[[15]] <- getGenesets(org = "spr", db = "kegg")
KEGG_Gene_Lists[[16]] <- getGenesets(org = "ssr", db = "kegg")
KEGG_Gene_Lists[[17]] <- getGenesets(org = "ssa", db = "kegg")
KEGG_Gene_Lists[[18]] <- getGenesets(org = "axy", db = "kegg")
KEGG_Gene_Lists[[19]] <- getGenesets(org = "bmj", db = "kegg")
KEGG_Gene_Lists[[20]] <- getGenesets(org = "mab", db = "kegg")


names(KEGG_Gene_Lists) <- Supported_Species$Species

#Save the KEGG IDs first so we can interface with the KEGG API
KEGG_Gene_Lists_IDs <- KEGG_Gene_Lists

for (i in 1:length(KEGG_Gene_Lists)) { #For all species
    names(KEGG_Gene_Lists[[i]]) <- substr(gsub("_", " ", str_extract(names(KEGG_Gene_Lists[[i]]), "[_]\\s*[A-Za-z0-9,_-]+\\s*")), 2, nchar(gsub("_", " ", str_extract(names(KEGG_Gene_Lists[[i]]), "[_]\\s*[A-Za-z0-9,_-]+\\s*"))))
}

for (i in 1:length(KEGG_Gene_Lists_IDs)) {
    names(KEGG_Gene_Lists_IDs[[i]]) <- substr(names(KEGG_Gene_Lists_IDs[[i]]), 1, 8)
}

Pathway_Name_Matching <- vector(mode = "list", length = nrow(Supported_Species))

for (i in 1:length(Pathway_Name_Matching)) {
  Pathway_Name_Matching[[i]] <- data.frame(names(KEGG_Gene_Lists[[i]]), names(KEGG_Gene_Lists_IDs[[i]]))
}
names(Pathway_Name_Matching) <- Supported_Species$Species

```

# 5. KEGG_Pathway_Data and KEGG_Pathway_List: Subset IntaRNA data for each species into individual KEGG pathways
```{r}
KEGG_Pathway_Data <- vector(mode = "list", length = nrow(Supported_Species))

#First, merge master annotation file into TargetPrediction Data and create KEGG_Pathway_Data
for (i in 1:length(KEGG_Pathway_Data)) {
  KEGG_Pathway_Data[[i]] <- merge(Target_Prediction_Data[[i]], Gene_Annotation_Data[[i]][[3]], by = "Gene_ID", all = TRUE)
}

names(KEGG_Pathway_Data) <- Supported_Species$Species
View(KEGG_Pathway_Data)

#Next, merge KEGG_Pathway_Data with KEGG_Gene_Lists to create KEGG_Pathway_List
KEGG_Pathway_List <- vector(mode = "list", length = nrow(Supported_Species))

for (j in 1:length(KEGG_Gene_Lists[[1]])) {
    KEGG_Pathway_List[[1]][[j]] <- KEGG_Pathway_Data[[1]][KEGG_Pathway_Data[[1]]$Gene_ID %in% KEGG_Gene_Lists[[1]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[2]])) {
  KEGG_Pathway_Data[[2]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[2]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[2]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[2]])) {
    KEGG_Pathway_List[[2]][[j]] <- KEGG_Pathway_Data[[2]][KEGG_Pathway_Data[[2]]$KEGG_ID %in% KEGG_Gene_Lists[[2]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[3]])) {
  KEGG_Pathway_Data[[3]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[3]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[3]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[3]])) {
    KEGG_Pathway_List[[3]][[j]] <- KEGG_Pathway_Data[[3]][KEGG_Pathway_Data[[3]]$KEGG_ID %in% KEGG_Gene_Lists[[3]][[j]],]
}

for (j in 1:length(KEGG_Gene_Lists[[4]])) {
    KEGG_Pathway_List[[4]][[j]] <- KEGG_Pathway_Data[[4]][KEGG_Pathway_Data[[4]]$Gene_ID %in% KEGG_Gene_Lists[[4]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[5]])) {
  KEGG_Pathway_Data[[5]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[5]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[5]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[5]])) {
    KEGG_Pathway_List[[5]][[j]] <- KEGG_Pathway_Data[[5]][KEGG_Pathway_Data[[5]]$KEGG_ID %in% KEGG_Gene_Lists[[5]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[6]])) {
  KEGG_Pathway_Data[[6]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[6]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[6]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[6]])) {
    KEGG_Pathway_List[[6]][[j]] <- KEGG_Pathway_Data[[6]][KEGG_Pathway_Data[[6]]$KEGG_ID %in% KEGG_Gene_Lists[[6]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[7]])) { 
  KEGG_Pathway_Data[[7]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[7]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[7]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[7]])) {
    KEGG_Pathway_List[[7]][[j]] <- KEGG_Pathway_Data[[7]][KEGG_Pathway_Data[[7]]$KEGG_ID %in% KEGG_Gene_Lists[[7]][[j]],]
}

View(KEGG_Pathway_Data[[7]])

for (i in 1:nrow(KEGG_Pathway_Data[[8]])) {
  KEGG_Pathway_Data[[8]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[8]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[8]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[8]])) {
    KEGG_Pathway_List[[8]][[j]] <- KEGG_Pathway_Data[[8]][KEGG_Pathway_Data[[8]]$KEGG_ID %in% KEGG_Gene_Lists[[8]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[9]])) {
  KEGG_Pathway_Data[[9]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[9]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[9]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[9]])) {
    KEGG_Pathway_List[[9]][[j]] <- KEGG_Pathway_Data[[9]][KEGG_Pathway_Data[[9]]$KEGG_ID %in% KEGG_Gene_Lists[[9]][[j]],]
}

for (j in 1:length(KEGG_Gene_Lists[[10]])) {
    KEGG_Pathway_List[[10]][[j]] <- KEGG_Pathway_Data[[10]][KEGG_Pathway_Data[[10]]$Gene_ID %in% KEGG_Gene_Lists[[10]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[11]])) {
  KEGG_Pathway_Data[[11]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[11]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[11]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[11]])) {
    KEGG_Pathway_List[[11]][[j]] <- KEGG_Pathway_Data[[11]][KEGG_Pathway_Data[[11]]$KEGG_ID %in% KEGG_Gene_Lists[[11]][[j]],]
}


for (j in 1:length(KEGG_Gene_Lists[[12]])) {
    KEGG_Pathway_List[[12]][[j]] <- KEGG_Pathway_Data[[12]][KEGG_Pathway_Data[[12]]$Gene_ID %in% KEGG_Gene_Lists[[12]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[13]])) {
  KEGG_Pathway_Data[[13]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[13]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[13]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[13]])) {
    KEGG_Pathway_List[[13]][[j]] <- KEGG_Pathway_Data[[13]][KEGG_Pathway_Data[[13]]$KEGG_ID %in% KEGG_Gene_Lists[[13]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[14]])) {
  KEGG_Pathway_Data[[14]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[14]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[14]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[14]])) {
    KEGG_Pathway_List[[14]][[j]] <- KEGG_Pathway_Data[[14]][KEGG_Pathway_Data[[14]]$KEGG_ID %in% KEGG_Gene_Lists[[14]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[15]])) {
  KEGG_Pathway_Data[[15]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[15]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[15]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[15]])) {
    KEGG_Pathway_List[[15]][[j]] <- KEGG_Pathway_Data[[15]][KEGG_Pathway_Data[[15]]$KEGG_ID %in% KEGG_Gene_Lists[[15]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[16]])) {
  KEGG_Pathway_Data[[16]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[16]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[16]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[16]])) {
    KEGG_Pathway_List[[16]][[j]] <- KEGG_Pathway_Data[[16]][KEGG_Pathway_Data[[16]]$KEGG_ID %in% KEGG_Gene_Lists[[16]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[17]])) {
  KEGG_Pathway_Data[[17]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[17]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[17]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[17]])) {
    KEGG_Pathway_List[[17]][[j]] <- KEGG_Pathway_Data[[17]][KEGG_Pathway_Data[[17]]$KEGG_ID %in% KEGG_Gene_Lists[[17]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[18]])) {
  KEGG_Pathway_Data[[18]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[18]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[18]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[18]])) {
    KEGG_Pathway_List[[18]][[j]] <- KEGG_Pathway_Data[[18]][KEGG_Pathway_Data[[18]]$KEGG_ID %in% KEGG_Gene_Lists[[18]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[19]])) {
  KEGG_Pathway_Data[[19]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[19]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[19]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[19]])) {
    KEGG_Pathway_List[[19]][[j]] <- KEGG_Pathway_Data[[19]][KEGG_Pathway_Data[[19]]$KEGG_ID %in% KEGG_Gene_Lists[[19]][[j]],]
}

for (i in 1:nrow(KEGG_Pathway_Data[[20]])) {
  KEGG_Pathway_Data[[20]]$KEGG_ID[i] <- substr(KEGG_Pathway_Data[[20]]$KEGG_ID[i], 5, nchar(KEGG_Pathway_Data[[20]]$KEGG_ID[i]))
}
for (j in 1:length(KEGG_Gene_Lists[[20]])) {
    KEGG_Pathway_List[[20]][[j]] <- KEGG_Pathway_Data[[20]][KEGG_Pathway_Data[[20]]$KEGG_ID %in% KEGG_Gene_Lists[[20]][[j]],]
}




for(i in 1:length(KEGG_Pathway_List)) {
  names(KEGG_Pathway_List[[i]]) <- names(KEGG_Gene_Lists[[i]])
}
names(KEGG_Pathway_List) <- Supported_Species$Species

View(KEGG_Pathway_List)


```

# 6. Target_Cutoff_Data: Further subsets KEGG_Pathway_List based on different targeting cutoffs
```{r}
#What is a proper targeting cutoff: We have a table of proteins that are significantly suppressed by a let-7b expressing strain of PA14 compared to a control strain. We can set the cutoffs in the app to be the Q1, median, and Q3 of the energy score distribution for these 48 genes (Table S4 here: https://www.pnas.org/doi/10.1073/pnas.2105370118#supplementary-materials) 

Energy_Scores <- c(-16.1, -19.2, -15.4, -15.4, -15.3, -13.1, -17.5, -11.0, -17.7, -13.8, -14.1, -15.3, -15.9, -11.6, -11.0, -16.1, -15.0, -15.6, -13.5, -20.2, -20.1, -16.8, -14.2, -14.5, -11.9, -19.2, -22.7, -12.2, -14.6, -16.1, -17.7, -15.4, -13.4, -14.6, -18.2, -13.2, -19.7, -15.8, -18.5, -10.1, -17.9, -16.1, -16.3, -13.8, -12.9, -13.4, -12.8, -14.6)

length(Energy_Scores) #48
summary(Energy_Scores) #Q1 = -16.98, Median = -15.35, Q3 = -13.47

Cutoffs <- c(-16.98, -15.35, -13.47)

#First set the targeting cutoffs
{
  Target_Cutoffs <- vector(mode = "list", length = nrow(Supported_Species))

  for (i in 1:length(Target_Cutoffs)) {
    for (j in 1:3) {
      Target_Cutoffs[[i]][[j]] <- Cutoffs[j]
    }
  }

  names(Target_Cutoffs) <- Supported_Species$Species
  for (i in 1:length(Target_Cutoffs)) {
    names(Target_Cutoffs[[i]]) <- c("Strong", "Medium", "Weak")
  }
  
  View(Target_Cutoffs)

}

#Need a list with five sub-objects per species - each with calculations based on the five different target cutoffs
{
#Target_Cutoff_List <- vector(mode = "list", length = nrow(Supported_Species))

for (a in 1:length(Target_Cutoff_List)) { 
  
  for (b in 1:3) { #Have to run the calculation for each targeting cutoff
  
    Chosen_Species <- KEGG_Pathway_List[[a]]   #Select each species one by one

    for (i in 1:length(Chosen_Species)) { #Go through all the pathways in each chosen species
    
    Chosen_Species[[i]][is.na(Chosen_Species[[i]])] <- 0 #First set all NAs to zeros to make the calculation possible
    
    if (nrow(Chosen_Species[[i]]) > 0) {
      
      for (j in 1:nrow(Chosen_Species[[i]])) { #Now run through each miRNA for the pathway and see if each gene meets the cutoff
      for (k in 2:631) {
        if (Chosen_Species[[i]][j,k] <= Target_Cutoffs[[a]][[b]]) { 
          Chosen_Species[[i]][j,k] <- 1
        } else {
          Chosen_Species[[i]][j,k] <- 0
        }
      }
    }
    
    } else {
      Chosen_Species[[i]] <- Chosen_Species[[i]]
    }
    
    
    #To see how things are running
    print(names(Chosen_Species)[i])
    print(b)
    print(i)
  }
  
  #Now assign the result to the Target_Cutoff_List object
  Target_Cutoff_List[[a]][[b]] <- Chosen_Species
  
  }
  
  
} 
  
View(Target_Cutoff_List)
  
###

names(Target_Cutoff_List) <- Supported_Species$Species
for(i in 1:length(Target_Cutoff_List)) {
  names(Target_Cutoff_List[[i]]) <- c("Strong", "Medium", "Weak")
}

}


#Make sure pathway names are right
for (i in 1:length(Target_Cutoff_List)) { 
  for (j in 1:length(Target_Cutoff_List[[i]])) { 
    names(Target_Cutoff_List[[i]][[j]]) <- names(KEGG_Pathway_List[[i]])
  }
} 

View(Target_Cutoff_List)

#Get rid of species specific tags to make names of pathways similar across all species
#for (i in 1:length(Target_Cutoff_List)) { #For all species
  #for (j in 1:length(Target_Cutoff_List[[i]])) { #For all five targeting cutoffs
    #names(Target_Cutoff_List[[i]][[j]]) <- substr(gsub("_", " ", str_extract(names(Target_Cutoff_List[[i]][[j]]), "[_]\\s*[A-Za-z0-9,_-]+\\s*")), 2, nchar(gsub("_", " ", str_extract(names(Target_Cutoff_List[[i]][[j]]), "[_]\\s*[A-Za-z0-9,_-]+\\s*"))))
 # }
#}



###

#Target_Cutoff_List_COPY <- Target_Cutoff_List #Create a copy in case I accidentally run the code above again and wipe out what I have created
#Target_Cutoff_List_COPY_10_31_22 <- Target_Cutoff_List
#Target_Cutoff_List_COPY_10_22 <- Target_Cutoff_List
#Target_Cutoff_List_COPY_10_31_22_2 <- Target_Cutoff_List
#Target_Cutoff_List_COPY_11_9_22 <- Target_Cutoff_List
#Target_Cutoff_List_COPY_12_2_22 <- Target_Cutoff_List
#Target_Cutoff_List_COPY_1_19_23 <- Target_Cutoff_List


#Before loading into the app, want to filter Target Cutoff List to remove rows for the same gene that has multiple UniProt IDs
Target_Cutoff_List_Unique <- Target_Cutoff_List

for (i in 1:length(Target_Cutoff_List_Unique)) {
  for (j in 1:length(Target_Cutoff_List_Unique[[i]])) {
    for (k in 1:length(Target_Cutoff_List_Unique[[i]][[j]])) {
      Target_Cutoff_List_Unique[[i]][[j]][[k]] <- Target_Cutoff_List_Unique[[i]][[j]][[k]][!duplicated(Target_Cutoff_List_Unique[[i]][[j]][[k]][ ,c("Gene_ID")]), ]
    }
  }
  print("done")
}

View(Target_Cutoff_List_Unique)

```




# 7. miRNA_Sequence_Data: Maps human miRNAs to their nucleotide sequences

First, load in the miRNA sequence data
```{r}
Raw_Sequence_Data <- read.table("./App_Data/human_mature_miRNA.txt")
```

Now clean up the data object so that it can be used in the app
```{r}
View(Raw_Sequence_Data) #Just need to put every other row in the name, sequence columns

miRNA_Names <- c()
miRNA_Sequences <- c()

for (i in 1:nrow(Raw_Sequence_Data)) {
  if (i %% 2 != 0) {
    miRNA_Names <- append(miRNA_Names, Raw_Sequence_Data[i,1])
  }
}

for (i in 1:nrow(Raw_Sequence_Data)) {
  if (i %% 2 == 0) {
    miRNA_Sequences <- append(miRNA_Sequences, Raw_Sequence_Data[i,1])
  }
}

miRNA_Sequence_Data <- data.frame(miRNA_Names, miRNA_Sequences)
colnames(miRNA_Sequence_Data) <- c("miRNA_Names", "miRNA_Sequences")

#Now just get rid of the '>' in all of the miRNA names
for (i in 1:nrow(miRNA_Sequence_Data)) {
  miRNA_Sequence_Data[i,1] <- substr(miRNA_Sequence_Data[i,1], 2, nchar(miRNA_Sequence_Data[i,1]))
}

#And replace '-' with '.' to match the format of miRNA names in the app
miRNA_Sequence_Data$miRNA_Names <- str_replace_all(miRNA_Sequence_Data$miRNA_Names, "-", ".")

View(miRNA_Sequence_Data)

#Now replace Ts with Us (NCBI and miRGeneDB - the databases where we have gathered our sequence data for this application - use T's instead of U's to represent the CDS and miRNA sequences. U (uracil) is the nucleotide that shows up in RNA and not T (thymine), so this is what should be shown in the application. Us are more biologically relevant ) 

for (i in 1:length(miRNA_Sequence_Data$miRNA_Sequences)) {
  miRNA_Sequence_Data$miRNA_Sequences[i] <- gsub("T", "U", miRNA_Sequence_Data$miRNA_Sequences[i])
}

View(miRNA_Sequence_Data)

```




# 8. Target_Protein_Data: Find the top miRNA that targets each protein
```{r}
Target_Protein_Data <- vector(mode = "list", length = nrow(Supported_Species))

for (a in 1:length(Target_Prediction_Data)) { 
  
  Species <- Target_Prediction_Data[[a]][,c(1:630)]
  Top_miRNAs <- vector(mode = "list")
  Top_Scores <- apply(Species, 1, min, na.rm = TRUE)
  
  for (i in 1:nrow(Species)) {
    Row_Select <- Species[i,]
    DF <- data.frame(colnames(Row_Select), unlist(Row_Select))
    miRName <- rownames(DF[which(DF$unlist.Row_Select. == Top_Scores[[i]]),])
    Top_miRNAs[[i]] <- miRName
    Top_miRNAs[[i]] <- toString(Top_miRNAs[[i]])
  }
  
  Protein_Description <- c(rep(0, nrow(Species)))
  
  Top_Proteins_DF <- data.frame(Protein_Description,
                                unlist(apply(Species, 1, min, na.rm = TRUE)),
                                unlist(Top_miRNAs))
  
  colnames(Top_Proteins_DF) <- c("Protein Description", "Best Energy Score", "Top Targeting miRNA(s)")

  for (i in 1:nrow(Top_Proteins_DF)) {
    Top_Proteins_DF$`Protein Description`[i] <- paste0(i, ". ", substr(str_extract(rownames(Top_Proteins_DF)[i], "protein=\\s*[A-Za-z0-9 ,()-/']+\\s*"), 9, nchar(str_extract(rownames(Top_Proteins_DF)[i], "protein=\\s*[A-Za-z0-9 ,()-/']+\\s*"))))
  }
  
  for (i in 1:nrow(Top_Proteins_DF)) {
    rownames(Top_Proteins_DF)[i] <- paste0(i, ". ", substr(str_extract(rownames(Top_Proteins_DF)[i], "locus_tag=\\s*[A-Za-z0-9_]+\\s*"), 11, nchar(str_extract(rownames(Top_Proteins_DF)[i], "locus_tag=\\s*[A-Za-z0-9_]+\\s*"))))
  }
  
  Target_Protein_Data[[a]] <- Top_Proteins_DF
    
  print("done")

}

names(Target_Protein_Data) <- Supported_Species$Species

View(Target_Protein_Data)

```


# 9. Energy Heat Maps: Identify top miRNAs by mean energy score and create data objects for heat map display
```{r}
Energy_HM_Data <- vector(mode = "list", length = nrow(Supported_Species))

for (i in 1:length(Energy_HM_Data)) {
  I <- Target_Prediction_Data[[i]][,c(1:630)]
  I[is.na(I)] <- 0
  Q <- quantile(apply(I ,2, mean), probs = seq(0, 1, 0.1), )[[2]]
  N <- names(apply(I ,2,mean)[which(apply(I ,2,mean) <= Q)])[1:63] 
  #Sometimes the miRNAs at the bottom of the list can tie so the top 10% is slightly larger for some species. Want to avoid this possibility - keep to the     #top 63 miRNAs in all cases even if we cut off one at the end of the list.
  Energy_HM_Data[[i]] <- Target_Prediction_Data[[i]][,which(colnames(Target_Prediction_Data[[i]]) %in% N)]
  Energy_HM_Data[[i]]$Protein_Names <- "NA"
  for (j in 1:nrow(Energy_HM_Data[[i]])) {
    Energy_HM_Data[[i]]$Protein_Names[j] <- substr(str_extract(rownames(Energy_HM_Data[[i]])[j], "protein=\\s*[A-Za-z0-9 ,()-/]+\\s*"), 9, nchar(str_extract(rownames(Energy_HM_Data[[i]])[j], "protein=\\s*[A-Za-z0-9 ,()-/]+\\s*")))
  }
}
  
names(Energy_HM_Data) <- Supported_Species$Species

```



# 10. Create .RData objects containing all data to load into app
```{r}
#These are the data files that are running on the public instance of the application
save(Supported_Species, file = "Data1.Rdata")
save(Target_Prediction_Data, file = "Data2.Rdata")
save(KEGG_Pathway_List, KEGG_Gene_Lists, file = "Data3.Rdata")
save(Pathway_Name_Matching, Target_Cutoffs, Target_Cutoff_List, file = "Data4.Rdata")
save(miRNA_Sequence_Data, Target_Protein_Data, Energy_HM_Data, file = "Data5.Rdata")

#If you are downloading the main instance from the Github repo, the data objects were created like this. It was necessary to break them up into smaller pieces in order to upload them to Github
TP1 <- Target_Prediction_Data[1:6]
TP2 <- Target_Prediction_Data[7:13]
TP3 <- Target_Prediction_Data[14:20]
save(TP1, file = "Data2a.Rdata")
save(TP2, file = "Data2b.Rdata")
save(TP3, file = "Data2c.Rdata")

KP1 <- KEGG_Pathway_List[1:9]
KP2 <- KEGG_Pathway_List[10:20]
save(KP1, file = "Data3a.Rdata")
save(KP2, KEGG_Gene_Lists, file = "Data3b.Rdata")

#In order to run the app, stitch the 3 objects together to form the single objects, Data2.Rdata and Data3.Rdata (This will improve app performance over loading in the three objects separately)
load("Data2a.Rdata")
load("Data2b.Rdata")
load("Data2c.Rdata")
Target_Prediction_Data <- c(TP1, TP2, TP3)
save(Target_Prediction_Data, file = "Data2.Rdata")

load("Data3a.Rdata")
load("Data3b.Rdata")
KEGG_Pathway_List <- c(KP1, KP2)
save(Target_Prediction_Data, KEGG_Gene_Lists, file = "Data3.Rdata")

View(KEGG_Pathway_List)


```



